fix(sudo): block all fw_setenv options, not only -s/--script (#208)
omnect/omnect-device-service: project-context.md

- `testfiles/` — test fixtures (positive/negative cases, Wi-Fi commissioning configs)
- `tests/` — integration tests for things with no Rust module to live in, e.g. the shell wrappers in `sudo/`

## 4. Repository-Specific Constraints

- **Mutually exclusive feature flags:** Exactly one of `bootloader_grub`, `bootloader_uboot`, or `mock` must be active. The `mock` feature stubs out hardware/systemd interactions for testing.
- **Feature trait pattern:** Every device capability implements the `Feature` trait (in its own `src/twin/*.rs` module). Adding a new feature means: (1) implement `Feature`, (2) add a `Command` variant, (3) register in `Twin::new()` feature map.
- **Command dispatch:** All operations flow through the `Command` enum and parsing helpers in `src/twin/feature/command.rs`, with file-watch handling in `fs_watcher.rs`. Direct methods, desired properties, file-system events, and intervals all produce `Command` values that get routed to the owning feature via `TypeId`.
- **Web service publish pattern:** Features publish state via `web_service::publish(PublishChannel, value)`. External consumers register endpoints in `/run/omnect-device-service/publish_endpoints.json`.

> - **Test location:** Unit tests live in a `#[cfg(test)] mod`, either inline in the file under test or in a sibling `*_test.rs` pulled in with `#[path]`. `tests/` is only for what has no Rust module to attach to, such as the shell wrappers in `sudo/`.

Reviewer: "either … or" reads as a free choice, and it is 15 to 1 - `src/twin/mod_test.rs` is the only `*_test.rs` in the tree. That invites new sibling files, which is the opposite of what you want the entry to do. Name the rule and the exception instead:

```suggestion
- **Test location:** Unit tests live in an inline `#[cfg(test)] mod tests` in the file under test. `src/twin/mod.rs` is the one exception and keeps them in a sibling `mod_test.rs` via `#[path]`. `tests/` is only for what has no Rust module to attach to, such as the shell wrappers in `sudo/`.
```

That also lines up with the IoT Hub mock entry right below, which already points at `src/twin/mod_test.rs`.

Author: Applied verbatim in `7b05a21`. Counted first: 15 files with an inline `mod tests`, and `src/twin/mod_test.rs` is the only `*_test.rs` — 15 to 1, as you said. "either … or" did read as a free choice.

- `sudo/` — sudoers drop-in files granting the service permission for specific privileged commands
- `polkit/` — polkit rules for D-Bus permissions
- `testfiles/` — test fixtures (positive/negative cases, Wi-Fi commissioning configs)

> - `tests/` — integration tests for things with no Rust module to live in, e.g. the shell wrappers in `sudo/`; everything else is tested in a `#[cfg(test)] mod` inside the file under test

Reviewer: The entry does two jobs: it describes `tests/`, and it states the convention for everything else. Section 3 is a file list, conventions live in section 4 "Repository-Specific Constraints" next to the `#[cfg(test)]` IoT Hub mock entry. Suggest keeping this line to the directory and moving the second half:

```suggestion
- `tests/` — integration tests for things with no Rust module to live in, e.g. the shell wrappers in `sudo/`
```

and in section 4:

```markdown
- **Test location:** Unit tests live in a `#[cfg(test)] mod` inside the file under test. `tests/` is only for what has no Rust module to attach to, such as the shell wrappers in `sudo/`.
```

Author: Applied both halves verbatim in `a6d29b2`. The section 4 entry sits directly above the `#[cfg(test)]` IoT Hub mock one, since both are about tests.

Author: Follow-up: I had to deviate from your suggestion after all, in `cd0919e`. Not the placement — that stays as you wrote it — but the first sentence.

"Unit tests live in a `#[cfg(test)] mod` inside the file under test" does not hold for every unit test here. `src/twin/mod.rs:5-7`:

```rust
#[cfg(test)]
#[path = "mod_test.rs"]
mod mod_test;
```

So `src/twin/mod.rs` keeps its tests in a sibling file — which is the same `src/twin/mod_test.rs` the IoT Hub mock entry two lines below points at. 16 other modules do use the inline form, so it is one exception, not a split convention. Current wording:

```markdown
- **Test location:** Unit tests live in a `#[cfg(test)] mod`, either inline in the file under test or in a sibling `*_test.rs` pulled in with `#[path]`. `tests/` is only for what has no Rust module to attach to, such as the shell wrappers in `sudo/`.
```

Flagging it rather than changing your text quietly, since the point of the entry is to tell the next contributor where to put a test.

Reviewer: You are right to flag it, and the exception is real - `src/twin/mod.rs:5-7` is as you quote it, and `src/twin/mod_test.rs` is the only `*_test.rs` in the tree. Keeping the `#[path]` case out of the rule would have been wrong.

One thing left on the wording, which I put on the current line so it can carry a suggestion. Resolving this thread.

Signed-off-by: Joerg Zeidler <joerg.zeidler@conplement.de>
